test_quality_of_service/test/test_deadline.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_test_publisher.hpp"
+#include "test_quality_of_service/qos_test_subscriber.hpp"
+#include "test_quality_of_service/qos_utilities.hpp"
+
+using namespace std::chrono_literals;
+
+///// Test Deadline with a single subscriber node
+TEST_F(QosRclcppTestFixture, test_deadline_no_publisher) {
+  const std::chrono::milliseconds deadline_duration = 1s;
+  const std::chrono::milliseconds test_duration = 10500ms;
+  const int expected_number_of_deadline_callbacks = static_cast<int>(
+    test_duration / deadline_duration);
+
+  int total_number_of_subscriber_deadline_events = 0;
+  int last_sub_count = 0;
+
+  // define qos profile
+  rclcpp::QoS qos_profile(10);
+  qos_profile.deadline(deadline_duration);
+
+  // setup subscription options and callback
+  rclcpp::SubscriptionOptions subscriber_options;
+  subscriber_options.event_callbacks.deadline_callback =
+    [&last_sub_count,
+    &total_number_of_subscriber_deadline_events](rclcpp::QOSDeadlineRequestedInfo & event) -> void
+    {
+      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "QOSDeadlineRequestedInfo callback");
```

Comment:
Use `Node::get_logger()`. In this case, `subscriber->get_logger()` I think.

Reply:
Hmm, this callback is being created before `subscriber`, so in order to use `subscriber->get_logger()`, there will need to be some non-trivial changes.

Reply:
Move line 65-68 to line 42 and replace `rclcpp::get_logger("rclcpp")` by `subscriber->get_logger()`. You'll need to also capture `subscriber` by reference.

Reply:
Okay, but this is quite confusing because `subscriber` and `subscriber_options` will be circularly referencing each other.

Reply:
You don't have to do it, but if you're having to bend over backwards to do this, then maybe you should rethink how you're organizing the code. Maybe the class that inherits from node shouldn't create the subscription in the constructor but by a method call. That way you could create the node, create the callback referencing the node, and then pass the options to a method on the node that creates the subscription.

Reply:
I'm becoming familiar with the code. I see a way to do it now. I'll update the PR when ready.

Reply:
Done. Can you please re-review, @wjwwood?